tutorials/haskell/golden-testing-in-haskell-with-hspec-gold/tutorial.md

```diff
@@ -0,0 +1,190 @@
+---
+title: Golden testing in Haskell with hspec-golden
+published: 2019-08-10
+tags: haskell, golden testing, snapshot testing, expect testing
+libraries: hspec-golen-0.1.0.0
+language: haskell
+author: Adrian Vera
+author-name: Adrian Vera
+twitter-profile: adrian_med
+github-profile: medAdrian
+description: In this tutorial we will implement golden tests in Haskell using hspec-golden a super set of the known test framework hspec. A golden test is an Input/Output action that records the result of a test into a file known as the "Golden File".
+---
+
+Golden testing (also known as Golden master testing or Snapshot testing) is pretty similar to unit testing, except that the test output is stored in a separate file known as the "Golden File".
+
+This kind of testing will help us verify modifications made to a reference version of our SUT(Subject Under Test) and validate that it has not modified its
+behaviour in an incorrect way. Thus giving us some safety when extending and refactoring code that doesn't have proper tests.
+
+When we write golden tests we build up knowledge of what the code is doing, this is very useful when we want to re-write it,
+we can run our test to find out immediately whether we've modified the desired behaviour (we start to see our tests in a different way
+since the purpose of golden testing is to document your system's behaviour, not to check for the wished one).
+
+Once we adopt this perspective, we have a completely different frame for understanding automated tests in general,
+we can see tests as descriptions of what we have rather than statements of correctness.
+We can review our tests periodically to tighten up their conditions as we decide what the behavior should be at each level of our systems.
+
+##The Main idea
+
+####‘A Golden File is your source of truth. You validate your test response against the Golden File. To summarize, the Golden file contains the response you expect from your test’
+Source: [Medium](https://medium.com/@jarifibrahim/golden-files-why-you-should-use-them-47087ec994bf) 
+
+##Changes on test flow
+The workflow for running the tests now becomes the following:
```

Comment:
@MedAdrian I know the workflow changes compared with the one described in TDD: 1) Write test, 2) Write specific amount of code, 3) Refactor it. Do you think it is worth to mention that, before listing the workflow for the golden tests? I'm thinking about something like:

```
In Test Driven Development , we follow the workflow:
1) Write a test first and make it fail
2) Write specific amount of code
3) Refactor it

However, in golden testing the workflow changes to:
1) Write a test. It will **succeed** at first time
2) Make changes in your SUT. The tests will fail
3) Update the golden file with the actual output of the SUT
```

It's a thought. I'm sure you can come up with something much better. 
Let me know what you think. :slightly_smiling_face:

Reply:
Sure it seems, i was wrong on my test flow, i've updated this on the new commit.